### DISCUSSION TITLE
Random PicklingError when retrieving data

### DISCUSSION THREAD

**Discussion by michaelbukachi** [sqlalchemy/sqlalchemy#8000] [Usage Questions]
I've been getting this error randomly when trying to cache models with their relationships. I'm using eager loading for all relationships (a mix of `joinedload` and `selectinload`). I have no idea where to start on debugging this. I'm not sure if this is a dogpile issue or a sqlalchemy issue :sweat_smile:. I'm using the snippets from the [examples](https://github.com/sqlalchemy/sqlalchemy/tree/main/examples/dogpile_caching) as a guide.
```
Can't pickle <function __init__ at 0x7f0bf74491f0>: it's not the same object as sqlalchemy.orm.instrumentation.__init__> for key
```
This is being caused by:
```python
      try:
          cached_value = dogpile_region.get_or_create(
              our_cache_key,
              createfunc,
              expiration_time=opt.expiration_time,
          )
      # TODO: Investigate what is causing these errors
      except (AttributeError, PicklingError) as e:
```
**Update**

On doing further tests, it seems this is being caused by `selectinload` loader

**Comment by zzzeek** [comment]
well it's definitely not a dogpile issue.   if you can narrow it down to selectinload then there may be something on there that needs better pickle reduction.   otherwise I would ensure all your mapped classes are created at the module level, things like that.   not sure what would cause it to want to pickle an `__init__` method which seems to be the one on a mapped class.

**Reply by michaelbukachi** [reply]
Did a test by stripping out a lot of the code. And did a test with just a dao method call while using `selectinload`, no pickling error. This is really weird. It seems there's something with my FastAPI configuration that is missing up with the setup :thinking: . All the mapped classes are module level.

**Reply by zzzeek** [reply]
pickle is very difficult to work with, because it's impossible to see what object it's screwing up on.   see if you can use pure python pickle and pdb in there to see what object it's not able to pickle.

**Reply by michaelbukachi** [reply]
Do you mean directly pickling the frozen result of the query returned?

**Reply by zzzeek** [reply]
maybe.  you have to get in there somehow to see what object it can't pickle

**Reply by michaelbukachi** [reply]
Using `dill` to debug. When running a trace on a frozen result it gives `cannot pickle 'KeyedRef' object`.

**Reply by zzzeek** [reply]
a KeyedRef.  lets find that...

**Reply by zzzeek** [reply]
that name looks famliar but I can't find it.  Can you peek at what `__module__` says on that class ?

**Reply by michaelbukachi** [reply]
On it.

**Reply by zzzeek** [reply]
also what SQLAlchemy version is this

**Reply by michaelbukachi** [reply]
It's part of the weakref module.

![screenshot_20220509175910](https://user-images.githubusercontent.com/10145850/167438273-e38cf99f-3c41-4c78-be7f-feeb243c1fc2.png)

**Reply by zzzeek** [reply]
i dont see how that would be related to "Can't pickle <function __init__ at 0x7f0bf74491f0>: it's not the same object as sqlalchemy.orm.instrumentation.__init__> for key"

**Reply by michaelbukachi** [reply]
Hmm. You are right. I'm still figuring out dill's error handling capabilities.
 Is there a way to reverse reference the memory address to a function/module in python? If I can get what `0x7f0bf74491f0` is pointing at then I can find out what module/function is causing the error.

**Reply by zzzeek** [reply]
the only way I do these are with pdb , I'm inside of plain python pickle, then I look at the objects and use "up" to walk up the stack to see what everything is.

**Reply by michaelbukachi** [reply]
So I've narrowed it to an insert statement on a particular model. Every time, an insert is done on this model it causes an error. The funny thing is that it causes whether a commit is made or not i.e the moment `db.execute(stmt)` is executed on the said model.

**Reply by zzzeek** [reply]
this brings #7861 to mind.  can you see if your error still occurs on version 1.4.32?

**Reply by zzzeek** [reply]
also is executing this insert part of an overall stack trace that includes the pickle issue?

**Reply by michaelbukachi** [reply]
We are running  1.4.32 already. The insert doesn't throw any error nor is it part of a stacktrace. The error is caused when querying later. The "problematic" model is being eager loaded as part of another model. It is also interesting to not that switching from `selectinload` to `joinedload` does remove the pickling error.

**Reply by zzzeek** [reply]
are you using declarative mapping or imperative?    the `__init__` method looks like it is pickling a class or an object before it is mapped, or something like that.   `sqlalchemy.orm.instrumentation.__init__` looks like the `__init__` method that is applied during the mapping process.

**Reply by michaelbukachi** [reply]
Declarative mapping. Here's the base class all models inherit:
```
BaseClass = declarative_base()


class Base(BaseClass, FilterSortMixin, InspectionMixin):
    __abstract__ = True

    id: Mapped[str] = Column(String, primary_key=True, default=generate_uuid)
    created_at: Mapped[datetime] = Column(
        DateTime, default=get_current_datetime, nullable=False
    )
    updated_at = Column(
        DateTime, default=None, onupdate=get_current_datetime, nullable=True
    )
```

**Reply by zzzeek** [reply]
OK then I got nothing on this end.   you'd need to get me a short reproducer

**Reply by michaelbukachi** [reply]
So I've been playing around with the query. Enabling/Disabling different parts of the query. So disabling a couple of joins from the query seems to resolve the issue. Does this mean this could be related to the size of the query? I'm seeing from the examples there's [RelationshipCache](https://github.com/sqlalchemy/sqlalchemy/blob/7a4f8770fc981837ac5f8ae8ed3a18f371f5e4d5/examples/dogpile_caching/caching_query.py#L155). However, it's only being used when lazy loading. Can it also be used for eager loading? :thinking:  Maybe, cache different parts of the query separately?

**Reply by zzzeek** [reply]
if you are referring to ``select().join()``, an explicit join, that wouldnt really be affecting anything except to the degree that it impacts what objects are being loaded and are therefore possibly not subject to being loaded by eager / lazy loaders.

maybe if the query itself is getting pickled as part of a selectinloader.   that's a possibility.

**Reply by zzzeek** [reply]
but....the "selectinloader" strategy itself, that's not pickled.   so...that makes no sense.

**Reply by michaelbukachi** [reply]
Following up on the suggestion from [here](https://github.com/sqlalchemy/sqlalchemy/discussions/8006#discussioncomment-2730052). I've tried pickling each col for every row from `frozen_result.rewrite_rows()`. I get PicklingError on some objs, When I run `print(dill.detect.baditems(col))`. The output is the obj itself i.e `[<....Model> object at memory>]`. Does this mean there's an issue with the instantiation of the relationship object?

**Reply by zzzeek** [reply]
we would assume here the pickling problem is due to some state on an instance of one of your objects, yes.  Keep in mind there's a whole "_sa_instance_state" collection on it that has a bunch of state in it.     But the error message originally shown here still looks like something related to the initial instrumentation of the class.

**Reply by zzzeek** [reply]
do you create any loader option objects ahead of time, like a `Load(Model).selectinload(Model.foo)` object that's at the model level which you reuse, something like that?

**Reply by michaelbukachi** [reply]
We initialize our load options this way:
```
some_dao = SomeDao(
    SomeObj,
    load_options=[
        joinedload(SomeObj.a).joinedload(A.b).joinedload(B.c)
    ],
)
```
This is a module-level declaration. We don't use `Load` in the codebase.

**Reply by zzzeek** [reply]
OK when you call that, can you call configure_mappers() first before you create those options?

**Reply by michaelbukachi** [reply]
Are you referring to adding a call to `configure_mappers` before instantiation of the daos?

**Reply by zzzeek** [reply]
yes...though I'm not sure if those joinedload() calls would be working if it weren't called....also i just am not sure how pickle would be getting confused in this way...but, something to try.   or to maybe create those options on the fly to see if that changes the behavior.

**Reply by zzzeek** [reply]
we are looking for mappers to be configured before joinedload(SomeObj.a) is called.

**Reply by michaelbukachi** [reply]
I can confirm that `configure_mappers` is called before the daos are instantiated. I've done a trace of one the models pickle is not able to handle using `dill`, here's the trace:
```
# T4
INFO:dill:# T4
T4: <class 'sqlalchemy.orm.instrumentation.ClassManager'>
INFO:dill:T4: <class 'sqlalchemy.orm.instrumentation.ClassManager'>
# T4
INFO:dill:# T4
F2: <function _queryable_attribute_unreduce at 0x7f2c884af790>
INFO:dill:F2: <function _queryable_attribute_unreduce at 0x7f2c884af790>
# F2
INFO:dill:# F2
D2: <dict object at 0x7f2c6c4310c0>
INFO:dill:D2: <dict object at 0x7f2c6c4310c0>
D2: <dict object at 0x7f2c6c431140>
INFO:dill:D2: <dict object at 0x7f2c6c431140>
# D2
INFO:dill:# D2
F1: <function __init__ at 0x7f2c6c432280>
INFO:dill:F1: <function __init__ at 0x7f2c6c432280>
F2: <function _create_function at 0x7f2c81c9fd30>
INFO:dill:F2: <function _create_function at 0x7f2c81c9fd30>
# F2
INFO:dill:# F2
Co: <code object __init__ at 0x7f2c6c437030, file "<string>", line 1>
INFO:dill:Co: <code object __init__ at 0x7f2c6c437030, file "<string>", line 1>
F2: <function _create_code at 0x7f2c81c9fdc0>
INFO:dill:F2: <function _create_code at 0x7f2c81c9fdc0>
# F2
INFO:dill:# F2
# Co
INFO:dill:# Co
D2: <dict object at 0x7f2c6c431c00>
INFO:dill:D2: <dict object at 0x7f2c6c431c00>
F1: <function _declarative_constructor at 0x7f2c88384af0>
INFO:dill:F1: <function _declarative_constructor at 0x7f2c88384af0>
Co: <code object _declarative_constructor at 0x7f2c88381d40, file "/home/michael/.cache/pypoetry/virtualenvs/app-d-AWDS2Y-py3.9/lib/python3.9/site-packages/sqlalchemy/orm/decl_base.py", line 1141>
INFO:dill:Co: <code object _declarative_constructor at 0x7f2c88381d40, file "/home/michael/.cache/pypoetry/virtualenvs/app-d-AWDS2Y-py3.9/lib/python3.9/site-packages/sqlalchemy/orm/decl_base.py", line 1141>
# Co
INFO:dill:# Co
D4: <dict object at 0x7f2c88368bc0>
INFO:dill:D4: <dict object at 0x7f2c88368bc0>
# D4
INFO:dill:# D4
D2: <dict object at 0x7f2c61433b00>
INFO:dill:D2: <dict object at 0x7f2c61433b00>
# D2
INFO:dill:# D2
# F1
INFO:dill:# F1
D2: <dict object at 0x7f2c6c435cc0>
INFO:dill:D2: <dict object at 0x7f2c6c435cc0>
# D2
INFO:dill:# D2
D4: <dict object at 0x7f2c8ad46500>
INFO:dill:D4: <dict object at 0x7f2c8ad46500>
# D4
INFO:dill:# D4
F1: <function __init__ at 0x7f2c6c432280>
INFO:dill:F1: <function __init__ at 0x7f2c6c432280>
D2: <dict object at 0x7f2c6c431d40>
INFO:dill:D2: <dict object at 0x7f2c6c431d40>
# D2
INFO:dill:# D2
# F1
INFO:dill:# F1
# D2
INFO:dill:# D2
# F1
INFO:dill:# F1
D2: <dict object at 0x7f2c6c431180>
INFO:dill:D2: <dict object at 0x7f2c6c431180>
# D2
INFO:dill:# D2
D2: <dict object at 0x7f2c6c431240>
INFO:dill:D2: <dict object at 0x7f2c6c431240>
T4: <class 'sqlalchemy.util.langhelpers.symbol'>
INFO:dill:T4: <class 'sqlalchemy.util.langhelpers.symbol'>
# T4
INFO:dill:# T4
# D2
INFO:dill:# D2
T4: <class 'sqlalchemy.event.base._UnpickleDispatch'>
INFO:dill:T4: <class 'sqlalchemy.event.base._UnpickleDispatch'>
# T4
INFO:dill:# T4
T4: <class 'sqlalchemy.orm.decl_api.registry'>
INFO:dill:T4: <class 'sqlalchemy.orm.decl_api.registry'>
# T4
INFO:dill:# T4
D2: <dict object at 0x7f2c832845c0>
INFO:dill:D2: <dict object at 0x7f2c832845c0>
T4: <class 'weakref.WeakValueDictionary'>
INFO:dill:T4: <class 'weakref.WeakValueDictionary'>
# T4
INFO:dill:# T4
D2: <dict object at 0x7f2c83284a80>
INFO:dill:D2: <dict object at 0x7f2c83284a80>
F1: <function WeakValueDictionary.__init__.<locals>.remove at 0x7f2c81de0ee0>
INFO:dill:F1: <function WeakValueDictionary.__init__.<locals>.remove at 0x7f2c81de0ee0>
Co: <code object remove at 0x7f2c8aa2c710, file "/usr/lib/python3.9/weakref.py", line 106>
INFO:dill:Co: <code object remove at 0x7f2c8aa2c710, file "/usr/lib/python3.9/weakref.py", line 106>
# Co
INFO:dill:# Co
D4: <dict object at 0x7f2c8aa0f640>
INFO:dill:D4: <dict object at 0x7f2c8aa0f640>
# D4
INFO:dill:# D4
R1: <weakref at 0x7f2c832529f0; to 'WeakValueDictionary' at 0x7f2c832b49d0>
INFO:dill:R1: <weakref at 0x7f2c832529f0; to 'WeakValueDictionary' at 0x7f2c832b49d0>
F2: <function _create_weakref at 0x7f2c81ca54c0>
INFO:dill:F2: <function _create_weakref at 0x7f2c81ca54c0>
# F2
INFO:dill:# F2
# R1
INFO:dill:# R1
B3: <built-in function _remove_dead_weakref>
INFO:dill:B3: <built-in function _remove_dead_weakref>
F2: <function _get_attr at 0x7f2c81ca5820>
INFO:dill:F2: <function _get_attr at 0x7f2c81ca5820>
# F2
INFO:dill:# F2
M2: <module '_weakref' (built-in)>
INFO:dill:M2: <module '_weakref' (built-in)>
F2: <function _import_module at 0x7f2c81ca5940>
INFO:dill:F2: <function _import_module at 0x7f2c81ca5940>
# F2
INFO:dill:# F2
# M2
INFO:dill:# M2
# B3
INFO:dill:# B3
D2: <dict object at 0x7f2c61409d40>
INFO:dill:D2: <dict object at 0x7f2c61409d40>
# D2
INFO:dill:# D2
# F1
INFO:dill:# F1
D2: <dict object at 0x7f2c8329ff80>
INFO:dill:D2: <dict object at 0x7f2c8329ff80>
```

I'm suspecting this line:
```
function WeakValueDictionary.__init__.<locals>.remove
```

is what is causing the :
```
TypeError: cannot pickle 'KeyedRef' object
```

**Reply by zzzeek** [reply]
here is the thing.  ClassManager should not be getting pickled at all.    It's an object associated with the class.  Here, we can see it's definitely not pickling all that:

```py
import pickle

from sqlalchemy import Column
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
Base = declarative_base()


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    data = Column(String)

print(pickle.dumps(A))
```

output:

```
b'\x80\x04\x95\x12\x00\x00\x00\x00\x00\x00\x00\x8c\x08__main__\x94\x8c\x01A\x94\x93\x94.'
```

the ClassManager has way more state than that.

**Reply by zzzeek** [reply]
this is why it seems like there's a class being created local to a scope somewhere, not at the module level.  something like that.

**Reply by zzzeek** [reply]
like _queryable_attribute_unreduce.  that callable is there to suit a case used by the "sqlalchemy.ext.serializer" object, which attempts (badly) to serialize queries.   it shouldn't be used.   There should be no way that callable is being part of a pickle for ORM mapped objects.    This might have to do with the mappings etc.

**Reply by michaelbukachi** [reply]
Hmm. We have a base class like so:
```python
BaseClass = declarative_base()


class Base(BaseClass, FilterSortMixin, InspectionMixin):
    __abstract__ = True

    id: Mapped[str] = Column(String, primary_key=True, default=generate_uuid)
    created_at: Mapped[datetime] = Column(
        DateTime, default=get_current_datetime, nullable=False
    )
    updated_at = Column(
        DateTime, default=None, onupdate=get_current_datetime, nullable=True
    )
```
Then all models inherit from it. All models are module-level. We do import all the models into a file called `base.py` but we only use it with alembic.

**Reply by zzzeek** [reply]
is `dill` the pickler you are using normally for your application?  I've not heard of this before and do you get the issue using normal Python pickle?

**Reply by michaelbukachi** [reply]
No. We are using `pickle`. I'm using `dill` for debugging since it has more tools for debugging pickling errors.

**Reply by zzzeek** [reply]
I've just run a test of pickling object structures loaded from a series of different loader strategy combinations:

```py
    _pathing_runs = [
        ("lazyload", "lazyload", "lazyload", 15),
        ("selectinload", "lazyload", "lazyload", 12),
        ("selectinload", "selectinload", "lazyload", 8),
        ("joinedload", "selectinload", "lazyload", 7),
        ("lazyload", "lazyload", "selectinload", 12),
        ("selectinload", "selectinload", "selectinload", 4),
        ("selectinload", "selectinload", "joinedload", 3),
    ]
```
no luck.   this is at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3833

**Reply by michaelbukachi** [reply]
Maybe I can give insights into our model organization:
We have:
```bash
db
|_base.py # Contains imports of all models from all feats. configure_mappers is also called here, after all models have been model.
|_base_class.py # Contains the `Base` class which is inherited by all models
feat_a
|_model.py
feat_b
|_model.py
```
All imports of models are from the declaration modules. I.e If there's a model `A` in `feat_a.model` then the import is `from feat.a import A`.
This also applies in the daos, where we are defining the load options. I have run the codebase using pdb and I can verify that all models are defined and `configure_mappers` is called before any of the daos are defined.

We do call `Base` from `db.base` but only in celery tasks so that all models are loaded in the background task.
